pkg/backup/processor.go

```diff
@@ -0,0 +1,561 @@
+/*
+Copyright 2017, 2019, 2020 the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backup
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"time"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/clock"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	velerov1client "github.com/vmware-tanzu/velero/pkg/generated/clientset/versioned/typed/velero/v1"
+	velerov1listers "github.com/vmware-tanzu/velero/pkg/generated/listers/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/label"
+	"github.com/vmware-tanzu/velero/pkg/metrics"
+	"github.com/vmware-tanzu/velero/pkg/persistence"
+	"github.com/vmware-tanzu/velero/pkg/plugin/clientmgmt"
+	"github.com/vmware-tanzu/velero/pkg/util/collections"
+	"github.com/vmware-tanzu/velero/pkg/util/encode"
+	kubeutil "github.com/vmware-tanzu/velero/pkg/util/kube"
+	"github.com/vmware-tanzu/velero/pkg/util/logging"
+	"github.com/vmware-tanzu/velero/pkg/volume"
+)
+
+// Processor handles end-to-end processing of a Backup request, from validation
+// and defaulting to execution & persistence to updating of the request object.
+type Processor struct {
```

Comment:
I like the `Processor` name, especially combined with the package name (`backup.Processor`, `restore.Processor` for the next one). The one thing I would nitpick about the name is the stutter in `Processor.Process`. Maybe `Processor.Run` or `Processor.Do`?

As far as alternatives for this type, just to strawman:

`Backupper` is already taken, so that doesn't fit here unless we want revisit renaming most of the package's high level types, which I'm not really wanting to do.

To get more specific. maybe `RequestProcessor`, since it only processes requests for new backups?

Or maybe shortening that, `Requester` and `Requester.Process`?

Reply:
Thanks for the input. I think I prefer `Processor.Run`, `RequestProcessor.Run`, or something along those lines. Agree re: Backupper - it might've made sense to name this type `Backupper`, and change the existing one to something else, but I don't really want to go down that path right now.

Reply:
@carlisia @ashish-amarnath any inputs on naming here?

Reply:
`Processor.Run()`  gets my vote

Reply:
I like making it more specific and explicit, so `RequestProcessor` is my favorite. `Processor` is not as clear imo. `Requester` is even less clear, in this context.

As for `RequestProcessor.Run`, the `.Run()` method does : "validates, defaults, runs and persists". It'd be nice to have a name that wasn't one of the actual thing it does, but I'm coming up empty. I'm not a fan of `.Do()`. Maybe `.Execute()`? I'd be good with `.Run()` tho, nbd.